### ISSUE TITLE
Node construction functions

### ISSUE THREAD

**Issue by michaelhkay** [qt4cg/qtspecs#573]
I propose introducing a set of functions that allow node construction in XPath. The basic functions are

```
new-document($children as node()*)
new-element($name as QName, $content as node()*)
new-attribute($name as QName, $value as xs:string)
new-namespace($prefix as xs:string, $uri as xs:string)
new-comment($content as xs:string)
new-processing-instruction($name as xs:string, $content as xs:string)
new-text($content as xs:string)
```

The semantics would be essentially identical to the current node constructors in XQuery and/or the corresponding instructions in XSLT (there are very few differences: a few minor ones, like if there are multiple attributes with the same name, XSLT takes the last, while XQuery throws an error).

As always, it's difficult to know where to draw the line in functionality between XPath and XQuery. One of the guidelines I use is that if an addition to XPath is likely to be useful to XSLT users, then it's worth including. Clearly these functions are not strictly necessary in XSLT (they could easily be user-supplied as wrappers around XSLT instructions). But to take advantage of some of the other capabilities we're introducing in XPath, node construction functions are increasingly handy. Consider for example:

`<xsl:copy-of select="interleave(para, new-element(xs:QName(xhtml:br)))"/>`

That's a one-liner that replaces

```
<xsl:for-each select="para">
  <xsl:if test="position() ne 1"><br/></xsl:if>
  <xsl:copy-of select="."/>
</xsl:for-each>
```

The new functions are also useful in XQuery because although they duplicate existing syntax, the fact that they are functions rather than custom syntax makes them more versatile.

As with existing constructs in XQuery and XSLT, a naive implementation that follows the semantics literally (which involves copying a subtree when adding an element to a new parent) would be rather inefficient. However, I think that the same established optimizations are equally applicable, for example lazy tree construction and/or push-mode evaluation.

**Comment by ndw** [comment]
One of the ways I (mentally) distinguish XPath from its host languages is that XPath can't construct things. (That's not strictly true given extension functions, but extension functions are sort of a Pandora's box in any event.)

I'm mildly opposed to adding node creation to XPath. But then, I don't actually find the function syntax in your example easier to read or understand than the XML syntax. More concise isn't always better.

**Comment by ChristianGruen** [comment]
If we decide that XPath should be capable of creating nodes, wouldn’t it be better to move the existing XQuery node construction features to the XPath specification?

Otherwise, we’d introduce yet another way to create nodes in XQuery – and it’s already confusing to some people that we have direct and computed node constructors.

**Comment by benibela** [comment]
So the issue is that XSLT users want to use features of XQuery, but do not have any XQuery syntax available?

Perhaps a simpler solution would be to allow using XQuery everywhere in XSLT where one can only use XPath now

>Consider for example:

>     <xsl:copy-of select="interleave(para, new-element(xs:QName(xhtml:br)))"/>

does this not look better:
  
     <xsl:copy-of select="interleave(para, element xhtml:br {} )"/>

     <xsl:copy-of select="interleave(para, &lt;xhtml:br/>)"/>
 

>If we decide that XPath should be capable of creating nodes, wouldn’t it be better to move the existing XQuery node construction features to the XPath specification?

and eventually you have moved everything from XQuery into XPath

There won't be any XPath left

**Comment by ChristianGruen** [comment]
> There won't be any XPath left

Ironically, to most users I talk to, XPath is XPath 1.0 anyway. Are there any XPath 3.x implementations at all that don't also support XQuery or XSLT?

**Comment by benibela** [comment]
> Are there any XPath 3.x implementations at all that don't also support XQuery or XSLT?

Not anymore? It used to be my goal. Implement XPath, and only XPath, because I do not need XQuery, so it would be  a leaner and more efficient implementation. But now I have implemented XQuery anyways. Because the test suite is mixing XPath and XQuery so much.

**Comment by michaelhkay** [comment]
>Perhaps a simpler solution would be to allow using XQuery everywhere in XSLT where one can only use XPath now

We could certainly consider expanding XPath to include everything in XQuery except the Prolog. There's a certain logic to that because the Prolog is essentially there to define a static context, and the typical role of XPath as a hosted expression language means that you want multiple expressions to share a single context.

My main reluctance to that is that it means bringing in some things that are completely superfluous in an XSLT environment. I guess I have only myself to blame by starting the ball rolling with node constructors, which can already be done in XSLT; but in that case the semantics of what I'm proposing is very close to the existing semantics of XSLT, and the syntax (using functions) is a lot more economical than the XQuery syntax which is completely custom-designed. Generally I would like to deliver as much functionality as possible using functions, because that's the most versatile and economical way of doing it.

**Comment by michaelhkay** [comment]
This seems not to have picked up much support from others, so I shall close it.

**Comment by ebruchez** [comment]
I realize that this is closed, but to me, the distinction between XPath and XQuery-without-prologue has always been more annoying than anything else. Case in point: no `let` in XPath 2 (then it came in XPath 3). The distinction has always been too arbitrary for my taste.

**Comment by michaelhkay** [comment]
There was an interesting discussion on the subject in connection with Mary Holstege's talk at Balisage today. She essentially has a requirement to maintain a function library that is usable from both XSLT and XQuery. This requirement tends to lead towards adding functionality to XPath that is currently available only in XSLT and XQuery (but in different ways).

**Comment by michaelhkay** [comment]
Here's another example from some code I had to write today:

```
   <xsl:variable name="remainder" as="text()">
      <xsl:value-of select="substring-after(., '!!!')"/>
   <xsl:variable>
   <xsl:apply-templates select="$remainder"/>
```

How much neater it would be to write

`<xsl:apply-templates select="text-node(substring-after(., '!!!'))"/>`

**Comment by michaelhkay** [comment]
I've just discovered that I prepared a draft specification for this which can be found at

https://www.saxonica.com/~mike/builder.html

**Comment by dnovatchev** [comment]
> One of the ways I (mentally) distinguish XPath from its host languages is that XPath can't construct things. (That's not strictly true given extension functions, but extension functions are sort of a Pandora's box in any event.)

@ndw It is actually possible - even in XPath 3.0.

(**_[hint](https://www.w3.org/TR/xpath-functions-30/#func-parse-xml)_**) 😄

**Comment by dnovatchev** [comment]
> I propose introducing a set of functions that allow node construction in XPath. The basic functions are
> 
> ```
> new-document($children as node()*)
> new-element($name as QName, $content as node()*)
> new-attribute($name as QName, $value as xs:string)
> new-namespace($prefix as xs:string, $uri as xs:string)
> new-comment($content as xs:string)
> new-processing-instruction($name as xs:string, $content as xs:string)
> new-text($content as xs:string)
> ```
> 
> The semantics would be essentially identical to the current node constructors in XQuery

Also, as observed in #1029 , we do need functions that produce:

- an empty sequence (constructor function for the type `empty-sequence()` )
- an empty array
- an empty map

If we had the function **`fn:const($x, $y) {$x}`** then the above three functions are expressed as:

- `fn:const((), ?)`
-  `fn:const([], ?)`
- `fn:const(map{} ?)`

And just  to remind that a trick exists that is used in many programming languages: we can implement a **_copy constructor_**  (deep copy) of anything by round-robin serialization then deserialization.

**Comment by ndw** [comment]
Yes, of course `parse-xml()` creates nodes. ¯\\_(ツ)_/¯ I'm not going to object to adding functions to create nodes; I was just observing that the distinction was (at least at one time) valuable (at least to me).

**Comment by michaelhkay** [comment]
I propose closing this with no action on the grounds that:

* It may be difficult to achieve consensus that the feature is required
* Designing the feature will require a significant amount of work
* We need to start cutting back on our ambitions for 4.0.
* We can live without it.

**Comment by ndw** [comment]
At [meeting 095](https://qt4cg.org/meeting/minutes/2024/10-22.html), the CG agreed to close this issue with no further action.

**Comment by michaelhkay** [comment]
I'm re-opening, as I think that a usable function for doing deep update of tree structures will require functions for creating new nodes.

**Comment by michaelhkay** [comment]
Abandoned as it proved difficult to get consensus.
